Add first chapter of flow-by-example to docs

spartan0x117 opened the pull request:
<!--

CONTRIBUTORS GUIDE: https://github.com/grafana/agent/blob/main/docs/developer/contributing.md#updating-the-changelog

If this is your first PR or you have not contributed in a while, we recommend
taking the time to review the guide. It gives helpful instructions for
contributors around things like how to update the changelog.

-->

#### PR Description
Adds the first section of flow-by-example to the docs. Note I'm not super familiar with the metadata at the top of the docs, but when looking at the docs locally everything seemed reasonable.

#### Notes to the Reviewer
@clayton-cornell where would be the most appropriate place to include the [docker-compose](https://github.com/spartan0x117/flow-by-example/blob/main/docker-compose.yml) so users can easily use it?

#### PR Checklist

<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [X] Documentation added

clayton-cornell commented:
@spartan0x117 We can include the YAML in the same directory as the tutorial, or you can add it to an assets directory similar to what has been done here: https://github.com/grafana/agent/tree/main/docs/sources/flow/tutorials/assets

clayton-cornell commented:
We should move all the screenshots/diagrams to the right location: https://grafana.com/docs/writers-toolkit/write/image-guidelines/#where-to-store-media-assets

rfratto commented:
Do the existing tutorials still hold educational value? Should this fully supplant the existing tutorials?

clayton-cornell commented:
> Do the existing tutorials still hold educational value? Should this fully supplant the existing tutorials?

My take on it... they should be removed or re-worked into a more comprehensive tutorials.

clayton-cornell commented:
> If we go that route, should that be done in this PR or a separate one?

I'd say a separate PR... otherwise we overload this one with scope creep. :-)

clayton-cornell commented on docs/sources/flow/tutorials/flow-by-example/faq.md:
```diff
@@ -0,0 +1,35 @@
+
+## Who is this for?
+
+This set of tutorials contains a collection of examples that build on each other to demonstrate how to configure and use the [Grafana Agent](https://grafana.com/docs/agent/latest/) in [Flow mode](https://grafana.com/docs/agent/latest/flow/). It assumes you have a basic understanding of what the Agent is and telemetry collection in general. It also assumes a base level of familiarity with Promethe ...
```
```suggestion
This set of tutorials contains a collection of examples that build on each other to demonstrate how to configure and use [{{< param "PRODUCT_NAME" >}}][flow]. It assumes you have a basic understanding of what {{< param "PRODUCT_ROOT_NAME" >}} is and telemetry collection in general. It also assumes a base level of familiarity with Prometheus and PromQL, Loki and LogQL, and basic Grafana navigation. It assumes no knowledge of {{< param "PRODUCT_NAME" >}} or River concepts.

[flow]: https://grafana.com/docs/agent/latest/flow
```

clayton-cornell commented on docs/sources/flow/tutorials/flow-by-example/faq.md:
```diff
@@ -0,0 +1,35 @@
+- /docs/grafana-cloud/send-data/agent/flow/tutorials/flow-by-example/faq/
+canonical: https://grafana.com/docs/agent/latest/flow/tutorials/flow-by-example/faq/
+description: FAQ About Flow-by-Example Tutorials
+title: FAQ
```
```suggestion
title: Get started
```

Could we title this another way? FAQ is... something I'd like to avoid if we can.

clayton-cornell commented on docs/sources/flow/tutorials/flow-by-example/first-components-and-stdlib/index.md:
```diff
@@ -0,0 +1,219 @@
+
+## River basics
+
+[Flow Configuration Language]: {{< relref "../../../config-language" >}}
```
```suggestion
[Configuration Language]: {{< relref "../../../config-language" >}}
```

clayton-cornell commented on docs/sources/flow/tutorials/flow-by-example/first-components-and-stdlib/index.md:
```diff
@@ -0,0 +1,219 @@
+## River basics
+
+[Flow Configuration Language]: {{< relref "../../../config-language" >}}
+[Flow Configuration Language Concepts]: {{< relref "../../../concepts/configuration_language" >}}
```
```suggestion
[Configuration Language Concepts]: {{< relref "../../../concepts/configuration_language" >}}
```

spartan0x117 commented on docs/sources/flow/tutorials/flow-by-example/first-components-and-stdlib/index.md:
```diff
@@ -0,0 +1,219 @@
+
+# First Components and the Standard Library
+
+_This section covers the basics of the River language and the standard library. It then introduces a basic pipeline that collects metrics from the host and sends them to Prometheus._
```
Nope! This is just a holdover from personal styling when it was a separate GitHub repo :sweat_smile:

clayton-cornell replied:
There might be one or two more then that should be tweaked. :-)

spartan0x117 commented:
I made a few changes after more carefully going through the writer's toolkit, mostly as individual commits that we can back out, if needed:

- I used a `collapse` shortcode for the solution rather than a downloadable file. Someone could copy-paste it if they wanted, but this also allows them to just view it easily. (I should probably also do this with the `docker-compose.yaml`).
- I also changed the list to be a repetitive list, and for styling I pushed the paragraph to a new line.
- Switched Headline Case -> Sentence case wherever I saw it.
- Replaced `relref` links with full urls, per [the writer's toolkit on links](https://grafana.com/docs/writers-toolkit/write/links/#link-to-grafanacom-pages) and added the `<AGENT_VERSION>` to them.

clayton-cornell commented:
> * I used a `collapse` shortcode for the solution rather than a downloadable file.

There's also a (default available) Collapse/Expand code thing for code snippets in codefence markdown. Longer code blocks are automatically collpased so that codeblocks don't eat up huge amounts of screen space :-)

![image](https://github.com/grafana/agent/assets/131809008/044f930c-ea6d-4f7d-b807-6cf754f5d65b)

spartan0x117 commented:
Good to know! :smile: In this case I was aiming to hide it entirely unless a reader _wants_ to see a potential solution (and do so ideally without having to leave the page).

clayton-cornell commented:
@spartan0x117 Are we done here? I think the docs are good enough to publish as-is.. get the info out there and we can update as needed.

spartan0x117 commented:
Yeah I think this one should be good to go! I can merge on Monday and then work on adding the other tutorial sections :smiley:

clayton-cornell commented:
@spartan0x117 Merged it
